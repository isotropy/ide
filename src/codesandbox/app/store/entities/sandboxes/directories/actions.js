// @flow

export const RENAME_DIRECTORY = 'RENAME_DIRECTORY';
export const MOVE_DIRECTORY = 'MOVE_DIRECTORY';

export default {
  renameDirectory: (id: string, title: string) => ({
    type: RENAME_DIRECTORY,
    id,
    title,
  }),
  moveDirectory: (id: string, directoryShortid: string) => ({
    type: MOVE_DIRECTORY,
    id,
    directoryShortid,
  }),
};
