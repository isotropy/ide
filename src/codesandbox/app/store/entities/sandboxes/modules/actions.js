// @flow

export const RENAME_MODULE = 'RENAME_MODULE';
export const MOVE_MODULE = 'MOVE_MODULE';
export const SET_CODE = 'SET_CODE';
export const SET_MODULE_SYNCED = 'SET_MODULE_SYNCED';
export const SET_MODULE_ERROR = 'SET_MODULE_ERROR';
export const CLEAR_MODULE_ERRORS = 'CLEAR_MODULE_ERRORS';

export const ADD_MODULE_CORRECTION = 'ADD_MODULE_CORRECTION';

export default {
  renameModule: (id: string, title: string) => ({
    type: RENAME_MODULE,
    id,
    title,
  }),
  moveModule: (id: string, directoryShortid: string) => ({
    type: MOVE_MODULE,
    id,
    directoryShortid,
  }),
  setCode: (id: string, code: string, isNotSynced: boolean = true) => ({
    type: SET_CODE,
    id,
    code,
    isNotSynced,
  }),
  setModuleSynced: (id: string) => ({
    type: SET_MODULE_SYNCED,
    id,
  }),
  clearErrors: () => ({
    type: CLEAR_MODULE_ERRORS,
  }),
  setError: (
    id: string,
    {
      title,
      message,
      line,
      column,
    }: { title: string, message: string, line: ?number, column: ?number }
  ) => ({
    id,
    type: SET_MODULE_ERROR,
    error: {
      title,
      message,
      line,
      column,
    },
  }),

  addCorrection: (
    id: string,
    {
      message,
      line,
      column,
      severity,
      source,
    }: {
      message: string,
      line: number,
      column: number,
      severity: 'warning' | 'notice',
      source: ?string,
    }
  ) => ({
    id,
    type: ADD_MODULE_CORRECTION,
    correction: {
      message,
      line,
      column,
      severity,
      source,
    },
  }),
};
