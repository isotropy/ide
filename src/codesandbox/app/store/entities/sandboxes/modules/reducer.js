// @flow
import type { Module } from 'codesandbox/common/types';
import { isEqual, mapValues } from 'lodash';

import {
  RENAME_MODULE,
  MOVE_MODULE,
  SET_CODE,
  SET_MODULE_SYNCED,
  SET_MODULE_ERROR,
  ADD_MODULE_CORRECTION,
  CLEAR_MODULE_ERRORS,
} from './actions';

function moduleReducer(module: Module, action): Module {
  switch (action.type) {
    case RENAME_MODULE:
      return { ...module, title: action.title };
    case MOVE_MODULE:
      return { ...module, directoryShortid: action.directoryShortid };
    case SET_CODE:
      return { ...module, code: action.code, isNotSynced: true };
    case SET_MODULE_SYNCED:
      return { ...module, isNotSynced: false };
    case ADD_MODULE_CORRECTION:
      if (!isEqual(action.correction, module.corrections[0])) {
        return {
          ...module,
          corrections: [...module.corrections, action.correction],
        };
      }

      return module;
    case SET_MODULE_ERROR: {
      if (!isEqual(action.error, module.errors[0])) {
        return { ...module, errors: [...module.errors, action.error] };
      }

      return module;
    }
    default:
      return module;
  }
}

type State = {
  [id: string]: Module,
};

export default function reducer(
  state: State,
  action: { type: string, id: string, [key: string]: any }
): State {
  switch (action.type) {
    case RENAME_MODULE:
    case MOVE_MODULE:
    case SET_CODE:
    case SET_MODULE_SYNCED:
    case ADD_MODULE_CORRECTION:
    case SET_MODULE_ERROR: {
      const module = state[action.id];
      if (module) {
        return {
          ...state,
          [action.id]: moduleReducer(module, action),
        };
      }
      return state;
    }
    case CLEAR_MODULE_ERRORS: {
      return mapValues(state, m => ({ ...m, errors: [], corrections: [] }));
    }
    default:
      return state;
  }
}
