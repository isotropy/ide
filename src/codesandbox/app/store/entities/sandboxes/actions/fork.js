// @flow
import { values } from 'lodash';

import { push } from 'react-router-redux';

import type { Module, Directory } from 'codesandbox/common/types';
import { sandboxUrl } from 'codesandbox/app/utils/url-generator';

import { createAPIActions, doRequest } from '../../../api/actions';
import { singleSandboxSelector } from '../selectors';
import { modulesSelector } from '../modules/selectors';
import { normalizeResult } from '../../actions';
import moduleActions from '../modules/actions';
import sandboxEntity from '../entity';
import notificationActions from '../../../notifications/actions';

export const FORK_SANDBOX_API_ACTIONS = createAPIActions('SANDBOX', 'FORK');

/**
 * When you fork you get a 'copy' of the modules, these modules have the shortid
 * in common, so if we want to get the module that is equivalent we want to use
 * this
 */
export const getEquivalentModule = (module: Module, modules: Array<Module>) => {
  const newModule = modules.find(
    m => m.id !== module.id && m.shortid === module.shortid
  );

  return newModule;
};

/**
 * When you fork you get a 'copy' of the directories, these directories have the shortid
 * in common, so if we want to get the directory that is equivalent we want to use
 * this
 */
export const getEquivalentDirectory = (
  directory: Directory,
  directories: Array<Directory>
) => {
  const newDirectory = directories.find(
    d => d.id !== directory.id && d.shortid === directory.shortid
  );

  return newDirectory;
};

export const forkSandbox = (id: string) => async (
  dispatch: Function,
  getState: Function
) => {
  const { data } = await dispatch(
    doRequest(FORK_SANDBOX_API_ACTIONS, `sandboxes/${id}/fork`, {
      method: 'POST',
    })
  );

  const currentSandbox = singleSandboxSelector(getState(), { id });
  if (currentSandbox) {
    data.currentModule = currentSandbox.currentModule;
  }
  data.forked = true;

  // Save the unsaved modules
  const oldModules = currentSandbox.modules.map(
    x => modulesSelector(getState())[x]
  );
  await dispatch(normalizeResult(sandboxEntity, data));

  const modules = values(modulesSelector(getState()));
  // Set the code for the new modules from the old modules
  oldModules.filter(m => m.isNotSynced).forEach(m => {
    // Mark old modules as synced so there is no confirm when moving to new url
    dispatch(moduleActions.setModuleSynced(m.id));

    const newModule = getEquivalentModule(m, modules) || m;
    dispatch(moduleActions.setCode(newModule.id, m.code));
  });

  dispatch(push(sandboxUrl(data)));

  dispatch(notificationActions.addNotification('Forked sandbox!', 'success'));

  return data;
};

/**
 * Will fork the sandbox if the sandbox is not owned
 * @param {string} sandboxId
 */
export const maybeForkSandbox = (sandboxId: string) => async (
  dispatch: Function,
  getState: Function
) => {
  const sandbox = singleSandboxSelector(getState(), { id: sandboxId });
  if (sandbox.owned) {
    // We own the sandbox, so return the id and don't fork
    return sandbox.id;
  }

  const forkedSandbox = await dispatch(forkSandbox(sandbox.id));

  return forkedSandbox.id;
};
