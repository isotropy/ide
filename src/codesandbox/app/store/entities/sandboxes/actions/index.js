// @flow
import { push } from 'react-router-redux';

import { newSandboxUrl } from 'codesandbox/app/utils/url-generator';

import { createAPIActions, doRequest } from '../../../api/actions';
import { normalizeResult } from '../../actions';
import notificationActions from '../../../notifications/actions';
import entity from '../entity';
import { singleSandboxSelector } from '../selectors';
import { modulesSelector } from '../modules/selectors';
import { directoriesSelector } from '../directories/selectors';

import { maybeForkSandbox, forkSandbox } from './fork';
import fileActions from './files';
import { currentUserSelector } from '../../../user/selectors';

export const SINGLE_SANDBOX_API_ACTIONS = createAPIActions('SANDBOX', 'SINGLE');
export const CREATE_SANDBOX_API_ACTIONS = createAPIActions('SANDBOX', 'CREATE');
export const DELETE_SANDBOX_API_ACTIONS = createAPIActions('SANDBOX', 'DELETE');
export const UPDATE_SANDBOX_API_ACTIONS = createAPIActions('SANDBOX', 'UPDATE');

export const UPDATE_NPM_DEPENDENCY_ACTIONS = createAPIActions(
  'SANDBOX',
  'UPDATE_NPM_DEPENDENCY'
);
export const DELETE_NPM_DEPENDENCY_ACTIONS = createAPIActions(
  'SANDBOX',
  'DELETE_NPM_DEPENDENCY'
);
export const ADD_EXTERNAL_RESOURCE_ACTIONS = createAPIActions(
  'SANDBOX',
  'ADD_EXTERNAL_RESOURCE'
);
export const DELETE_EXTERNAL_RESOURCE_ACTIONS = createAPIActions(
  'SANDBOX',
  'DELETE_EXTERNAL_RESOURCE'
);
export const SET_SANDBOX_PRIVACY_ACTIONS = createAPIActions(
  'SANDBOX',
  'PRIVACY'
);
export const ADD_TAG_ACTIONS = createAPIActions('SANDBOX', 'ADD_TAG');
export const REMOVE_TAG_ACTIONS = createAPIActions('SANDBOX', 'REMOVE_TAG');
export const LIKE_SANDBOX_ACTIONS = createAPIActions('SANDBOX', 'LIKE');
export const UNLIKE_SANDBOX_ACTIONS = createAPIActions('SANDBOX', 'UNLIKE');
export const SET_SANDBOX_INFO = 'SET_SANDBOX_INFO';
export const SET_NPM_DEPENDENCIES = 'SET_NPM_DEPENDENCIES';
export const SET_EXTERNAL_RESOURCES = 'SET_EXTERNAL_RESOURCES';
export const SET_TAGS = 'SET_TAGS';
export const SET_CURRENT_MODULE = 'SET_CURRENT_MODULE';
export const SET_PROJECT_VIEW = 'SET_PROJECT_VIEW';
export const SET_VIEW_MODE = 'SET_VIEW_MODE';
export const CREATE_ZIP = 'CREATE_ZIP';
export const DEPLOY_SANDBOX = 'DEPLOY_SANDBOX';
export const SET_SANDBOX_PRIVACY = 'SET_SANDBOX_PRIVACY';
export const FORCE_RENDER = 'FORCE_RENDER';

export default {
  updateSandboxInfo: (id: string, title: string, description: string) => async (
    dispatch: Function,
    getState: Function
  ) => {
    const sandboxId = await dispatch(maybeForkSandbox(id));
    const {
      title: oldTitle,
      description: oldDescription,
    } = singleSandboxSelector(getState(), { id: sandboxId });

    dispatch({
      type: SET_SANDBOX_INFO,
      id: sandboxId,
      title,
      description,
    });
    try {
      await dispatch(
        doRequest(UPDATE_SANDBOX_API_ACTIONS, `sandboxes/${sandboxId}`, {
          body: {
            sandbox: { title, description },
          },
          method: 'PUT',
        })
      );
    } catch (e) {
      dispatch({
        type: SET_SANDBOX_INFO,
        id: sandboxId,
        title: oldTitle,
        description: oldDescription,
      });
    }
  },

  setViewMode: (id: string, showEditor: boolean, showPreview: boolean) => ({
    type: SET_VIEW_MODE,
    id,
    showEditor,
    showPreview,
  }),

  setProjectView: (id: string, isInProjectView: boolean) => ({
    type: SET_PROJECT_VIEW,
    id,
    isInProjectView,
  }),

  setCurrentModule: (id: string, moduleId: string, lineNumber?: number) => ({
    type: SET_CURRENT_MODULE,
    id,
    moduleId,
    lineNumber,
  }),

  getById: (id: string) => async (dispatch: Function) => {
    const { data } = await dispatch(
      doRequest(SINGLE_SANDBOX_API_ACTIONS, `sandboxes/${id}`)
    );

    dispatch(normalizeResult(entity, data));

    return data;
  },

  createSandbox: () => async (dispatch: Function) => {
    const { data } = await dispatch(
      doRequest(CREATE_SANDBOX_API_ACTIONS, `sandboxes`, {
        method: 'POST',
        body: { sandbox: {} },
      })
    );

    await dispatch(normalizeResult(entity, data));

    return data;
  },

  forkSandbox,

  addTag: (id: string, tag: string) => async (
    dispatch: Function,
    getState: Function
  ) => {
    const sandboxId = await dispatch(maybeForkSandbox(id));

    const sandbox = singleSandboxSelector(getState(), { id: sandboxId });

    // Already do change
    dispatch({
      type: SET_TAGS,
      id: sandboxId,
      tags: [...sandbox.tags, tag],
    });

    try {
      const result = await dispatch(
        doRequest(ADD_TAG_ACTIONS, `sandboxes/${sandboxId}/tags`, {
          method: 'POST',
          body: {
            tag,
          },
        })
      );
      dispatch({
        type: SET_TAGS,
        id: sandboxId,
        tags: result.data,
      });
    } catch (e) {
      dispatch({
        type: SET_TAGS,
        id: sandboxId,
        tags: sandbox.tags,
      });
      throw e;
    }
  },

  removeTag: (id: string, tag: string) => async (
    dispatch: Function,
    getState: Function
  ) => {
    const sandboxId = await dispatch(maybeForkSandbox(id));

    const sandbox = singleSandboxSelector(getState(), { id: sandboxId });

    // Already do change
    dispatch({
      type: SET_TAGS,
      id: sandboxId,
      tags: sandbox.tags.filter(t => t !== tag),
    });

    try {
      const result = await dispatch(
        doRequest(REMOVE_TAG_ACTIONS, `sandboxes/${sandboxId}/tags/${tag}`, {
          method: 'DELETE',
        })
      );
      dispatch({
        type: SET_TAGS,
        id: sandboxId,
        tags: result.data,
      });
    } catch (e) {
      dispatch({
        type: SET_TAGS,
        id: sandboxId,
        tags: sandbox.tags,
      });
      throw e;
    }
  },

  addNPMDependency: (id: string, dependency: string, version: string) => async (
    dispatch: Function
  ) => {
    const sandboxId = await dispatch(maybeForkSandbox(id));
    const result = await dispatch(
      doRequest(
        UPDATE_NPM_DEPENDENCY_ACTIONS,
        `sandboxes/${sandboxId}/dependencies`,
        {
          method: 'POST',
          body: {
            dependency: {
              name: dependency,
              version,
            },
          },
        }
      )
    );

    dispatch({
      type: SET_NPM_DEPENDENCIES,
      id: sandboxId,
      dependencies: result.data,
    });
  },

  removeNPMDependency: (id: string, dependency: string) => async (
    dispatch: Function
  ) => {
    const sandboxId = await dispatch(maybeForkSandbox(id));

    const result = await dispatch(
      doRequest(
        DELETE_NPM_DEPENDENCY_ACTIONS,
        `sandboxes/${sandboxId}/dependencies/${dependency}`,
        {
          method: 'DELETE',
        }
      )
    );

    dispatch({
      type: SET_NPM_DEPENDENCIES,
      id: sandboxId,
      dependencies: result.data,
    });
  },

  addExternalResource: (id: string, resource: string) => async (
    dispatch: Function
  ) => {
    const sandboxId = await dispatch(maybeForkSandbox(id));
    const result = await dispatch(
      doRequest(
        ADD_EXTERNAL_RESOURCE_ACTIONS,
        `sandboxes/${sandboxId}/resources`,
        {
          method: 'POST',
          body: {
            external_resource: resource,
          },
        }
      )
    );

    dispatch({
      type: SET_EXTERNAL_RESOURCES,
      id: sandboxId,
      externalResources: result.data,
    });
  },

  removeExternalResource: (id: string, resource: string) => async (
    dispatch: Function
  ) => {
    const sandboxId = await dispatch(maybeForkSandbox(id));
    const result = await dispatch(
      doRequest(
        DELETE_EXTERNAL_RESOURCE_ACTIONS,
        `sandboxes/${sandboxId}/resources/`,
        {
          method: 'DELETE',
          body: {
            id: resource,
          },
        }
      )
    );

    dispatch({
      type: SET_EXTERNAL_RESOURCES,
      id: sandboxId,
      externalResources: result.data,
    });
  },

  createZip: (id: string) => async (dispatch: Function, getState: Function) => {
    dispatch({
      type: CREATE_ZIP,
      id,
    });
    const modules = modulesSelector(getState());
    const directories = directoriesSelector(getState());
    const sandbox = singleSandboxSelector(getState(), { id });

    const createZip = await import('../utils/create-zip');

    createZip.default(
      sandbox,
      sandbox.modules.map(x => modules[x]),
      sandbox.directories.map(x => directories[x])
    );
  },

  deploy: (id: string) => async (dispatch: Function, getState: Function) => {
    dispatch({
      type: DEPLOY_SANDBOX,
      id,
    });
    const modules = modulesSelector(getState());
    const directories = directoriesSelector(getState());
    const sandbox = singleSandboxSelector(getState(), { id });

    const deploy = await import('../utils/deploy');

    const apiData = await deploy.default(
      sandbox,
      sandbox.modules.map(x => modules[x]),
      sandbox.directories.map(x => directories[x])
    );

    const user = currentUserSelector(getState());
    const token = user.integrations.zeit.token;

    let res;
    try {
      res = await fetch('https://api.zeit.co/now/deployments', {
        method: 'POST',
        body: JSON.stringify(apiData),
        headers: {
          Authorization: `bearer ${token}`,
        },
      });
    } catch (e) {
      dispatch(
        notificationActions.addNotification(
          'An unknown error occured when connecting to ZEIT',
          'error'
        )
      );

      throw e;
    }

    const data = await res.json();

    if (!res.ok) {
      dispatch(notificationActions.addNotification(data.err.message, 'error'));
      throw new Error(data.err.message);
    }

    return data.host && `https://${data.host}`;
  },

  deleteSandbox: (id: string) => async (dispatch: Function) => {
    await dispatch(
      doRequest(DELETE_SANDBOX_API_ACTIONS, `sandboxes/${id}`, {
        method: 'DELETE',
        body: {
          id,
        },
      })
    );

    dispatch(notificationActions.addNotification('Deleted Sandbox', 'success'));
  },

  newSandboxUrl: () => async (dispatch: Function) => {
    dispatch(push(newSandboxUrl()));
  },

  likeSandbox: (id: string) => async (dispatch: Function) => {
    await dispatch(
      doRequest(LIKE_SANDBOX_ACTIONS, `sandboxes/${id}/likes`, {
        method: 'POST',
        body: {
          id,
        },
      })
    );
  },

  unLikeSandbox: (id: string) => async (dispatch: Function) => {
    await dispatch(
      doRequest(UNLIKE_SANDBOX_ACTIONS, `sandboxes/${id}/likes`, {
        method: 'DELETE',
        body: {
          id,
        },
      })
    );
  },

  setSandboxPrivacy: (id: string, privacy: number) => async (
    dispatch: Function,
    getState: Function
  ) => {
    const oldPrivacy = singleSandboxSelector(getState(), { id });
    dispatch({
      type: SET_SANDBOX_PRIVACY,
      id,
      privacy,
    });

    try {
      const result = await dispatch(
        doRequest(SET_SANDBOX_PRIVACY_ACTIONS, `sandboxes/${id}/privacy`, {
          method: 'PATCH',
          body: {
            sandbox: {
              privacy,
            },
          },
        })
      );
      return result;
    } catch (e) {
      dispatch({ type: SET_SANDBOX_PRIVACY, id, privacy: oldPrivacy });

      throw e;
    }
  },

  forceRender: (id: string) => ({
    type: FORCE_RENDER,
    id,
  }),

  ...fileActions,
};
