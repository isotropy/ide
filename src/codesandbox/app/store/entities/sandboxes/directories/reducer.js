// @flow
import type { Directory } from 'codesandbox/common/types';
import { RENAME_DIRECTORY, MOVE_DIRECTORY } from './actions';

function directoryReducer(directory: Directory, action) {
  switch (action.type) {
    case RENAME_DIRECTORY:
      return { ...directory, title: action.title };
    case MOVE_DIRECTORY:
      return { ...directory, directoryShortid: action.directoryShortid };
    default:
      return directory;
  }
}

type State = {
  [id: string]: Directory,
};

export default function reducer(
  state: State,
  action: { type: string, [key: string]: any }
): State {
  switch (action.type) {
    case RENAME_DIRECTORY:
    case MOVE_DIRECTORY:
      if (state[action.id]) {
        return {
          ...state,
          [action.id]: directoryReducer(state[action.id], action),
        };
      }
      return state;
    default:
      return state;
  }
}
