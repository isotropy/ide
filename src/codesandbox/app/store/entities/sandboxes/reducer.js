// @flow
import type { Sandbox } from 'codesandbox/common/types';
import { mapValues } from 'lodash';

import {
  SET_NPM_DEPENDENCIES,
  SET_EXTERNAL_RESOURCES,
  SET_CURRENT_MODULE,
  SET_SANDBOX_INFO,
  SET_PROJECT_VIEW,
  SET_VIEW_MODE,
  DELETE_SANDBOX_API_ACTIONS,
  LIKE_SANDBOX_ACTIONS,
  UNLIKE_SANDBOX_ACTIONS,
  SET_TAGS,
  SET_SANDBOX_PRIVACY,
  FORCE_RENDER,
} from './actions';

import {
  ADD_MODULE_TO_SANDBOX,
  ADD_DIRECTORY_TO_SANDBOX,
  REMOVE_MODULE_FROM_SANDBOX,
  REMOVE_DIRECTORY_FROM_SANDBOX,
} from './actions/files';

import { SET_CURRENT_USER, SIGN_OUT } from '../../user/actions';

const initialState = {};

type Action = {
  [key: string]: any,
  type: string,
};

function singleSandboxReducer(sandbox: Sandbox, action: Action): Sandbox {
  switch (action.type) {
    case SET_VIEW_MODE:
      return {
        ...sandbox,
        showEditor: action.showEditor,
        showPreview: action.showPreview,
      };
    case SET_PROJECT_VIEW:
      return { ...sandbox, isInProjectView: action.isInProjectView };
    case SET_CURRENT_MODULE:
      return { ...sandbox, currentModule: action.moduleId };
    case ADD_MODULE_TO_SANDBOX:
      return { ...sandbox, modules: [...sandbox.modules, action.moduleId] };
    case ADD_DIRECTORY_TO_SANDBOX:
      return {
        ...sandbox,
        directories: [...sandbox.directories, action.directoryId],
      };
    case REMOVE_MODULE_FROM_SANDBOX: {
      const currentModule = sandbox.currentModule;
      const resetCurrentModule = currentModule === action.moduleId;

      return {
        ...sandbox,
        currentModule: resetCurrentModule ? undefined : sandbox.currentModule,
        modules: sandbox.modules.filter(m => m !== action.moduleId),
      };
    }
    case REMOVE_DIRECTORY_FROM_SANDBOX:
      return {
        ...sandbox,
        directories: sandbox.directories.filter(d => d !== action.directoryId),
      };
    case SET_NPM_DEPENDENCIES:
      return {
        ...sandbox,
        npmDependencies: action.dependencies,
        dependencyBundle: {}, // So we can fetch dependencies later on again
      };
    case SET_EXTERNAL_RESOURCES:
      return {
        ...sandbox,
        externalResources: action.externalResources,
      };
    case SET_SANDBOX_INFO:
      return {
        ...sandbox,
        title: action.title,
        description: action.description,
      };
    case LIKE_SANDBOX_ACTIONS.REQUEST:
      if (sandbox.userLiked) return sandbox;

      return {
        ...sandbox,
        likeCount: sandbox.likeCount + 1,
        userLiked: true,
      };
    case LIKE_SANDBOX_ACTIONS.SUCCESS:
      return {
        ...sandbox,
        likeCount: action.data.count,
      };
    case LIKE_SANDBOX_ACTIONS.FAILURE:
      return {
        ...sandbox,
        likeCount: sandbox.likeCount - 1,
        userLiked: false,
      };
    case UNLIKE_SANDBOX_ACTIONS.SUCCESS:
      return {
        ...sandbox,
        userLiked: false,
        likeCount: action.data.count,
      };
    case UNLIKE_SANDBOX_ACTIONS.FAILURE:
      return {
        ...sandbox,
        userLiked: true,
        likeCount: sandbox.likeCount + 1,
      };
    case UNLIKE_SANDBOX_ACTIONS.REQUEST:
      if (!sandbox.userLiked) return sandbox;

      return {
        ...sandbox,
        userLiked: false,
        likeCount: sandbox.likeCount - 1,
      };
    case SET_TAGS:
      return {
        ...sandbox,
        tags: action.tags,
      };
    case SET_SANDBOX_PRIVACY:
      return {
        ...sandbox,
        privacy: action.privacy,
      };
    case FORCE_RENDER:
      return {
        ...sandbox,
        forcedRenders: sandbox.forcedRenders + 1,
      };
    default:
      return sandbox;
  }
}

export default function reducer(
  state = initialState,
  action: Action
): { [id: string]: Sandbox } {
  switch (action.type) {
    case ADD_MODULE_TO_SANDBOX:
    case ADD_DIRECTORY_TO_SANDBOX:
    case REMOVE_MODULE_FROM_SANDBOX:
    case REMOVE_DIRECTORY_FROM_SANDBOX:
    case SET_NPM_DEPENDENCIES:
    case SET_EXTERNAL_RESOURCES:
    case SET_CURRENT_MODULE:
    case SET_SANDBOX_INFO:
    case SET_PROJECT_VIEW:
    case SET_VIEW_MODE:
    case SET_TAGS:
    case LIKE_SANDBOX_ACTIONS.REQUEST:
    case LIKE_SANDBOX_ACTIONS.SUCCESS:
    case LIKE_SANDBOX_ACTIONS.FAILURE:
    case UNLIKE_SANDBOX_ACTIONS.REQUEST:
    case UNLIKE_SANDBOX_ACTIONS.FAILURE:
    case UNLIKE_SANDBOX_ACTIONS.SUCCESS:
    case SET_SANDBOX_PRIVACY:
    case FORCE_RENDER: {
      const id = action.id || (action.meta ? action.meta.id : undefined);
      if (state[id]) {
        return {
          ...state,
          [id]: singleSandboxReducer(state[id], action),
        };
      }

      return state;
      // The user has changed, we need to mark all sandboxes as owned if the author
      // id corresponds with the new user id
    }
    case SET_CURRENT_USER:
      return mapValues(state, s => ({
        ...s,
        owned: s.owned || s.author === action.data.id,
      }));
    case SIGN_OUT:
      return mapValues(state, s => ({
        ...s,
        owned: false,
      }));
    case DELETE_SANDBOX_API_ACTIONS.SUCCESS:
      return {
        ...state,
        [action.meta.id]: null,
      };
    default:
      return state;
  }
}
