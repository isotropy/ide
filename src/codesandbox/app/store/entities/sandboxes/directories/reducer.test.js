import { createDirectory } from 'codesandbox/common/test/mocks';
import reducer from './reducer';
import actions from './actions';

describe('directoriesReducer', () => {
  it('can move a directory', () => {
    const directory = createDirectory();
    const state = { [directory.id]: directory };

    const newState = reducer(
      state,
      actions.moveDirectory(directory.id, '218hj')
    );

    expect(newState).toMatchSnapshot();
  });

  it('can rename a directory', () => {
    const directory = createDirectory();
    const state = { [directory.id]: directory };

    const newState = reducer(
      state,
      actions.renameDirectory(directory.id, 'ttest2')
    );

    expect(newState).toMatchSnapshot();
  });
});
