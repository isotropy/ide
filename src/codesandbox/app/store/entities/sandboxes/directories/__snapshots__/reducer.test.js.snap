// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`directoriesReducer can move a directory 1`] = `
Object {
  "longid-dir0": Object {
    "directoryShortid": "218hj",
    "id": "longid-dir0",
    "shortid": "shortid-dir0",
    "title": "test-dir0",
  },
}
`;

exports[`directoriesReducer can rename a directory 1`] = `
Object {
  "longid-dir0": Object {
    "directoryShortid": null,
    "id": "longid-dir0",
    "shortid": "shortid-dir0",
    "title": "ttest2",
  },
}
`;
